gotham/src/router/builder/single.rs

```diff
@@ -106,6 +109,57 @@ pub trait DefineSingleRoute {
     where
         H: Handler + RefUnwindSafe + Copy + Send + Sync + 'static;
 
+    /// Similar to `to`, but accepts an `async fn`
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # extern crate gotham;
+    /// # extern crate hyper;
+    /// #
+    /// # use hyper::{Body, Response, StatusCode};
+    /// # use gotham::handler::HandlerResult;
+    /// # use gotham::state::State;
+    /// # use gotham::router::Router;
+    /// # use gotham::router::builder::*;
+    /// # use gotham::pipeline::new_pipeline;
+    /// # use gotham::pipeline::single::*;
+    /// # use gotham::middleware::session::NewSessionMiddleware;
+    /// # use gotham::test::TestServer;
+    /// #
+    /// async fn my_handler(state: State) -> HandlerResult {
+    ///     // Handler implementation elided.
+    /// #   Ok((state, Response::builder().status(StatusCode::ACCEPTED).body(Body::empty()).unwrap()))
+    /// }
+    /// #
+    /// # fn router() -> Router {
+    /// #   let (chain, pipelines) = single_pipeline(
+    /// #       new_pipeline().add(NewSessionMiddleware::default()).build()
+    /// #   );
+    ///
+    /// build_router(chain, pipelines, |route| {
+    ///     route.get("/request/path").to_async(my_handler);
+    /// })
+    /// #
+    /// # }
+    /// #
+    /// # fn main() {
+    /// #   let test_server = TestServer::new(router()).unwrap();
+    /// #   let response = test_server.client()
+    /// #       .get("https://example.com/request/path")
+    /// #       .perform()
+    /// #       .unwrap();
+    /// #   assert_eq!(response.status(), StatusCode::ACCEPTED);
+    /// # }
+    /// ```
+    fn to_async<H, Fut>(self, handler: H)
+    where
+        Self: Sized,
+        H: (FnOnce(State) -> Fut) + RefUnwindSafe + Copy + Send + Sync + 'static,
+        Fut: Future<Output = HandlerResult> + Send + 'static,
+    {
+        self.to_new_handler(move || Ok(move |s: State| handler(s).boxed()))
+    }
```

Comment:
I have implemented this directly on the trait because it feels like the most backwards-compatible thing to do.

In practice, the only implementation of DefineSingleRoute in the gotham codebase is SingleRouteBuilder. Would you rather I implement it on SingleRouteBuilder and force all external implementors of DefineSingleRoute to implement it?

Also, can someone look extra-hard at the marker traits that I've used on H and Fut, and make sure they're all appropriate? (I copied some of them from .to(), and I'm not sure they're all needed)

Reply:
I'm not sure. In a world where you don't care about backwards compatibility, which would you go for? The versioning model we are on allows us to break things, and we can provide documentation for migration if it's complicated.

The traits seem ok to me (when looking at a diff). I feel like `Copy` and `'static` might be unnecessary...?

Reply:
Ideally, the `.to` should map to `(State, Response<Body>)` (normal fn), `Pin<Box<HandlerFuture>>` (future fn) and `HandlerResult` (async fn).

We can take it one step forward and disallow the mapping to `(State, Response<Body>)` and replace it with `HandlerResult` for consistency.

Reply:
If both future fn and async fn could not be supported at the same time and we are willing to break (since it is a major version anyway), then we should give priority to async fn and create a new `to_future` or something for the future fn.

Reply:
@whitfin that sounds like the right question to ask. Good thinking. In an ideal world, I would probably think about making this trait private or something (I think that a recent compiler release that allows traits to be marked nonexhaustive, so they can't be implemented by external people?). In the short term, "program to the interface" seems to be the pattern that's being used here, so refactoring the implementation down into SingleRouteBuilder sounds like the thing to do. I will add this to the checklist, and do it once I have all of the other questions worked out.



@pksunkara: (I think that this is a separate discussion) as it stands in my PR (using your terminology) I can pass a "future fn" to `.to_async()`, because Pin<Box<HandlerFuture>> implements `Future<Output = HandlerResult>`, but I can't pass "normal fn".

When I wrote this, I didn't think it's possible to create a single function that can accept both an `impl Trait` and a struct. I seem to remember, rustc complaining that my blanket `impl HandlerFuture` for `impl Future` might conflict with my `impl HandlerFuture` for `(State, Response<Body>)` if someone else implements `Future` for `(State, Response<Body>)`... or something. You can get away with it for `Pin<Box<HandlerFuture>>` because in that case your impl is over two struct types that the compiler can guarantee won't overlap. This was last year though, so I might be misremembering, or the compiler might have become more lenient since then.

I've been mulling it over in my head over the last couple of days, and I think that it might be possible to impl Future for `(State, Response<Body>)`, and get on with our lives. Do you think that would work?

Reply:
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=19c7034bdda2cb834f574c041788111f 
```
error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
 --> src/lib.rs:7:1
  |
7 | impl Future for (State, String) {
  | ^^^^^^^^^^^^^^^^---------------
  | |               |
  | |               this is not defined in the current crate because tuples are always foreign
  | impl doesn't use only types from inside the current crate
  |
  = note: define and implement a trait or new type instead
```
:(

Reply:
How about making the return object `HandlerResult` for `normal fn`

Reply:
same deal: HandlerResult is just a Result that's generic over tuples, so it' also foreign.
```
9 | impl Future for HandlerResult {
  | ^^^^^^^^^^^^^^^^-------------
  | |               |
  | |               `std::result::Result` is not defined in the current crate
  | impl doesn't use only types from inside the current crate
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=1e88ddbdf314a06d0871b2b7fe10ee82

I think that we need a separate `.to()` for tuples vs `.to_async()` for futures. It might be that we should call them `.to_sync()` and `.to()` though.

Reply:
> The traits seem ok to me (when looking at a diff). I feel like Copy and 'static might be unnecessary...?

I had a quick play about and I couldn't see a way to get rid of them.

The handler callback is `FnOnce + Copy`. This is a bit like `Fn`, but a bit easier to implement with closures or something? If I remove the `+ Copy` then I end up with a handler function that can only be run once.

`+ 'static` is required by `.to_new_handler()`. It feels a bit excessive at first glance, but I think it's fine: the routes we define normally live for the entire duration of the program, and are normally implemented as top-level `fn`s. I haven't really dug into whether there is a more appropriate lifetime than this, so I'm just going to keep what's there.